go/arrow/compute/datum.go

```diff
@@ -0,0 +1,346 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compute
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/array"
+	"github.com/apache/arrow/go/arrow/scalar"
+)
+
+//go:generate go run golang.org/x/tools/cmd/stringer -type=ValueShape -linecomment
+//go:generate go run golang.org/x/tools/cmd/stringer -type=DatumKind -linecomment
+
+// ValueShape is a brief description of the shape of a value (array, scalar or otherwise)
+type ValueShape int8
+
+const (
+	// either Array or Scalar
+	ShapeAny    ValueShape = iota // any
+	ShapeArray                    // array
+	ShapeScalar                   // scalar
+)
+
+// ValueDescr is a descriptor type giving both the shape and the datatype of a value
+// but without the data.
+type ValueDescr struct {
+	Shape ValueShape
+	Type  arrow.DataType
+}
+
+func (v *ValueDescr) String() string {
+	return fmt.Sprintf("%s [%s]", v.Shape, v.Type)
+}
+
+// DatumKind is an enum used for denoting which kind of type a datum is encapsulating
+type DatumKind int
+
+const (
+	KindNone       DatumKind = iota // none
```

Comment:
Just a side note, I'm not sure if all of these are actually used in the C++ implementation.